src/utils/cpu.js

```diff
@@ -0,0 +1,201 @@
+import getLowestEmptyYIndex from "./getLowestEmptyYIndex";
+
+const GameState = {
+  game: "GAME",
+  player_win: "PLAYER_WIN",
+  cpu_win: "CPU_WIN",
+  draw: "DRAW",
+};
+
+/**
+ * CPUのクラス
+ */
+class Cpu {
+  /**
+   * Cpuクラスのコンストラクターの説明
+   * @param {string[][]} board - 盤面を表す二次元配列
+   * @param {number} victoryCondition - 勝利条件
+   * @param {string} cpuName - boardで使用しているcpuの名前（Player2など）
+   * @param {string} playerName - boardで使用しているplayerの名前（Player1など）
+   */
+  constructor(board, victoryCondition, cpuName, playerName) {
+    this.board = board;
+    this.victoryCondition = victoryCondition;
+    this.cpuName = cpuName;
+    this.playerName = playerName;
+    this.myTurn = false; // false: CPU, true: Player
+    this.state = GameState.game;
+  }
+
+  /**
+   * ターンを交代する
+   */
+  changeTurn() {
+    this.myTurn = !this.myTurn;
+  }
+
+  /**
+   * 座標に対する操作を取り消す
+   * @param {number} x - 操作を取り消す座標x
+   * @param {number} y - 操作を取り消す座標y
+   */
+  undoValue(x, y) {
+    this.board[x][y] = null;
+    this.changeTurn();
+  }
+
+  /**
+   * 指定の座標に石を置く
+   * @param {number} x - 石を置く座標x
+   * @param {number} y - 石を置く座標y
+   */
+  putStone(x, y) {
+    this.myTurn ? (this.board[x][y] = this.playerName) : (this.board[x][y] = this.cpuName);
+    this.checkState(x, y);
+    this.changeTurn();
+  }
+
+  /**
+   * ゲームの状態を更新する
+   * @param {number} x - 直前に石を置いた座標x
+   * @param {number} y - 直前に石を置いた座標y
+   */
+  checkState(x, y) {
+    let winner = calculateWinner(this.board, this.victoryCondition, x, y);
+    if (winner === this.cpuName) {
+      this.state = GameState.cpu_win;
+    } else if (winner === this.playerName) {
+      this.state = GameState.player_win;
+    } else if (winner === "draw") {
+      this.state = GameState.draw;
+    } else {
+      this.state = GameState.game;
+    }
+  }
+
+  /**
+   * 評価関数
+   * @param {number} x - 直前に石を置いた座標x
+   * @param {number} y - 直前に石を置いた座標y
+   * @param {number} depth - ゲーム木の深さ
+   * @returns 評価値
+   */
+  evaluate(x, y, depth) {
+    let maxScore = this.board.length * this.board[0].length + 1;
+    let winner = calculateWinner(this.board, this.victoryCondition, x, y);
+    if (winner === this.cpuName) {
+      return maxScore - depth;
+    } else if (winner == this.playerName) {
+      return depth - maxScore;
+    } else {
+      return 0;
+    }
+  }
+
+  /**
+   * 次の手を考える
+   * @param {string} mode - easy/medium/hard
+   * @returns cpuが石を置く座標x
+   */
+  cpuThink(mode) {
+    let maxScore = this.board.length * this.board[0].length + 1;
+    if (mode === "easy") {
+      return this.random(); // ランダム
+    } else if (mode === "medium") {
+      return this.minmax(0, -maxScore, maxScore, null, null, 3); // 3手先まで読む
+    } else if (mode === "hard") {
+      return this.minmax(0, -maxScore, maxScore, null, null, 10); // 10手先まで読む
+    }
+  }
+
+  /**
+   * 石を置ける座標xをランダムに返す
+   * @returns 石を置く座標x
+   */
+  random() {
+    for (;;) {
+      let x = Cpu.getRandomInt(this.board.length);
```

Comment:
```suggestion
      const x = Cpu.getRandomInt(this.board.length);
```
`getRandomInt`だけ静的メソッドになっていますが、何か理由がありますか？単純に気になりました。

Reply:
理由としては、classのメンバーに関係のない関数だからですね。